python/lsst/daf/base/yaml.py

```diff
@@ -0,0 +1,111 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""This module defines YAML I/O for key lsst.daf.base classes."""
+
+# If yaml package is not installed there is no reason to fail everywhere
+try:
+    import yaml
+except ImportError:
+    yaml = None
+
+from .dateTime import DateTime
+from .propertyContainer import PropertyList, getPropertyListState, setPropertyListState, \
+    getPropertySetState, setPropertySetState, PropertySet
+
+
+# YAML representers for key lsst.daf.base classes
+
+
+def dt_representer(dumper, data):
+    """Represent an lsst.daf.base.DateTime (as ISO8601-formatted string in TAI)
+    """
+    return dumper.represent_scalar('lsst.daf.base.DateTime',
+                                   data.toString(DateTime.TAI))
+
+
+if yaml:
+    yaml.add_representer(DateTime, dt_representer)
+
+
+def pl_representer(dumper, data):
+    """Represent an lsst.daf.base.PropertyList as an ordered sequence of
+    name/type/value/comment lists)"""
+    # Turn the tuples into lists for cleaner representation in yaml
+    result = getPropertyListState(data, asLists=True)
+    return dumper.represent_sequence('lsst.daf.base.PropertyList', result,
+                                     flow_style=None)
+
+
+if yaml:
+    yaml.add_representer(PropertyList, pl_representer)
+
+
+def ps_representer(dumper, data):
+    """Represent an lsst.daf.base.PropertySet as a mapping from names to
+    type/value pairs."""
+    # Turn the tuples into lists for cleaner representation in yaml
+    result = getPropertySetState(data, asLists=True)
+    return dumper.represent_sequence('lsst.daf.base.PropertySet', result,
+                                     flow_style=None)
+
+
+if yaml:
+    yaml.add_representer(PropertySet, ps_representer)
+
+###############################################################################
+
+# YAML constructors for key lsst.daf.base classes
+
+
+def dt_constructor(loader, node):
+    """Construct an lsst.daf.base.DateTime from an ISO8601-formatted string in
+    TAI"""
+    dt = loader.construct_scalar(node)
+    return DateTime(str(dt), DateTime.TAI)
+
+
+if yaml:
+    yaml.add_constructor('lsst.daf.base.DateTime', dt_constructor)
+
+
+def pl_constructor(loader, node):
+    """Construct an lsst.daf.base.PropertyList from a YAML pickle-like structure."""
+    pl = PropertyList()
+    yield pl
+    state = loader.construct_sequence(node, deep=True)
+    setPropertyListState(pl, state)
+
+
+if yaml:
+    yaml.add_constructor('lsst.daf.base.PropertyList', pl_constructor)
+
+
+def ps_constructor(loader, node):
+    """Construct an lsst.daf.base.PropertyList from a YAML pickle-like structure."""
+    ps = PropertySet()
+    yield ps
+    state = loader.construct_sequence(node, deep=True)
+    setPropertySetState(ps, state)
```

Comment:
Will this correct handle a hierarchical PropertySet? How about a contained PropertyList? I'm concerned because PropertySet has several ways of iterating and I'm not sure which will be used.

Reply:
Nesting PropertySet and PropertyList is a bit odd. It works to a certain extent but it seems to be a bit broken since the hierarchy only manifests itself in the "."-delimited key names and not the actual types. If I put a list inside a set you get something like:
```yaml
!<lsst.daf.base.PropertySet>
- [newpl.new.bool, Bool, true]
- [newpl.new.withcom, String, string]
- [newpl.new.int, Int, 2008]
- [newpl.new.ps.bool, Bool, true]
- [newpl.new.ps.short, Short, 42]
- [newpl.new.ps.ps.string2, String, bar]
- [newpl.ps.ps.string2, String, bar]
- [newpl.new.ps.int, Int, 2008]
- [newpl.new.ps.ps.string, String, foo]
- [newpl.ps.ps.string, String, foo]
- [newpl.ps.int, Int, 2008]
- [newpl.ps.short, Short, 42]
- [newpl.ps.bool, Bool, true]
- [newpl.bool, Bool, true]
- [newpl.int, Int, 2008]
- [newpl.new.short, Short, 42]
- [newpl.withcom, String, string]
- [newpl.short, Short, 42]
- [bool, Bool, true]
- [short, Short, 42]
- [int, Int, 2008]
- [ps.string, String, foo]
- [ps.string2, String, bar]
```
if you put a set inside a list you get:
```yaml
!<lsst.daf.base.PropertyList>
- [bool, Bool, true, '']
- [short, Short, 42, '']
- [int, Int, 2008, '']
- [withcom, String, string, a comment]
- [ps.bool, Bool, true, '']
- [ps.short, Short, 42, '']
- [ps.int, Int, 2008, '']
- [ps.ps.string, String, foo, '']
- [ps.ps.string2, String, bar, '']
- [new.ps.ps.string2, String, bar, '']
- [new.ps.ps.string, String, foo, '']
- [new.ps.int, Int, 2008, '']
- [new.ps.short, Short, 42, '']
- [new.ps.bool, Bool, true, '']
- [new.int, Int, 2008, '']
- [new.withcom, String, string, a comment]
- [new.short, Short, 42, '']
- [new.bool, Bool, true, '']
```

so the original notion of set vs list is lost and if you put a list inside a set all the comments are stripped and if you put a set inside a list it looks like empty comments. Ideally you'd want the hierarchy to persist in the yaml file. I think pickling will have the same problem of information loss.  I haven't fully understood PropertyList yet because if I `PropertyList.setPropertySet()` and try to `getPropertySet` it doesn't work (I can't get the set back again). It works if I am doing `getPropertySet` on a `PropertySet`. I have to get the individual dot-delimited entries instead.

This uses the pickling infrastructure so it seems to me that pickling a List inside a Set would also lead to the comments disappearing.

I'll add the extra tests but I'm not sure I can fix the problem here. I think it would be better to try to fix it once I make things more dict-like with dunder methods. Then you really do want a get on a propertySet inside another propertySet to return a new dict-like thing and not simply all the keys with all the dots in them.